lib/CommandClient.php

```diff
@@ -26,22 +32,77 @@ public static function socketPath(string $config) {
     private function send($message): Promise {
         return call(function () use ($message) {
             if (!$this->socket) {
-                $this->socket = yield from $this->establish();
+                if (!$this->socketPromise) {
+                    new Coroutine(self::start($this->path, $this->socket, $this->socketPromise, $this->readDeferreds));
+                }
+
+                yield $this->socketPromise;
+            }
+
+            try {
+                $message = \json_encode($message) . "\n";
+                yield $this->socket->write($message);
+            } catch (ClosedException $e) {
+                $this->socket = null;
+                new Coroutine(self::start($this->path, $this->socket, $this->socketDeferred, $this->readDeferreds));
+                yield $this->socketPromise;
+                yield $this->socket->write($message);
             }
 
-            $message = \json_encode($message);
-            $message = \pack("N", \strlen($message)) . $message;
-            return yield $this->socket->write($message);
+            $promise = ($this->readDeferreds[] = new Deferred)->promise();
+
+
+            return $promise;
         });
     }
 
-    private function establish(): \Generator {
+    private static function start($path, &$socket, &$socketPromise, &$readDeferreds) {
```

Comment:
This function seems very awkward using things by reference. I'd like to suggest another approach that I can put together later today.